### ISSUE TITLE
Think outside the service box (just a little)

### ISSUE THREAD

**Issue by donaldquixote** [ratpack/ratpack#1109]
tl;dr
It would be nice to see more official support for using Ratpack constructs outside of a RatpackServer instance. For instance, I can create my own DefaultExecController today and do most of the things I need, but I'm pretty sure there was a reason why this class is located in the ratpack.exec._internal_ package.

The longer version:

My services provide command line entry points that start an application which is not a Ratpack service. I would like to be able to reuse code components between the command line application and the service application, including some Ratpack-based components (e.g. interfaces returning Promise). 

I realize that I could factor my code in a way so that shared components are Ratpack-agnostic and I then wrap those with Ratpack interfaces, but my early attempts at this have led to the realization that this severely limits my ability to take advantage of all Ratpack has to offer when using these components in the service application.

So when starting up my command line application, I instead instantiate a DefaultExecController with a thread count = 1, then fork an execution to perform the bulk of my work in. 

For the most part everything works great... 

It just feels dirty because I think it's not officially supported...

Oh, and if ExecStarter could provide a way to take a pre-built registry instead of forcing the RegistrySpec mechanism, that would be nice, too :)

Thanks.

**Comment by danveloper** [comment]
Hi @donaldquixote, in the few disparate talks that we've had about what a Ratpack 2.0 would look like, we've hit on the idea that things like the Execution Model, functional interfaces, and registry abstractions should be decoupled and usable outside of a web application. I think you're right in line with what we're thinking.

In terms of what you're doing now, by instantiating the `DefaultExecController` ... yeah, that's really the only way to do what you're looking for today. You're right though, it is unsupported, and because `DefaultExecController` is an `internal` construct, the API may break at any time without notice.

What we might consider in the nearer-than-2.0-term is providing some public API that would allow users to stand up an `Execution` on the current thread, without the need for the ELG & friends. @ratpack/committers wdyt about this idea? Might buy us some runway to be able to use Ratpack constructs outside of a web app without a complete 2.0-worthy refactor.

**Comment by donaldquixote** [comment]
Thanks for the reply. It's good to hear that this is being considered. Would the short term approach you suggest allow you to work with Promises and the Blocking construct still? I'm just seeking a bit of clarity or scoping around what "stand up an Execution" means.

Thanks!

**Comment by johnrengelman** [comment]
@danveloper why not just provide a public static method on `ExecController` that creates a new `DefaultExecController`? What's the downside of having the ELG & friends? If you don't, wouldn't constructs such as `Blocking.get()` break?

I think some type of stop gap would be warranted. Ratpack 2.x is probably still a good distance in the future and really just getting the dev a handle to `ExecController` should be sufficient for anything they need. It's ultimately up to them when it's appropriate to fork new executions and such, so I don't think there would be a great way to abstract around that.
